implementations/rust/ockam/ockam_command/src/upgrade.rs

```diff
@@ -1,51 +1,157 @@
+    pub fn run(self, opts: CommandGlobalOpts) {
+        local_cmd(run_impl(opts, self));
+    }
+}
+
+fn run_impl(opts: CommandGlobalOpts, cmd: UpgradeCommand) -> miette::Result<()> {
+    let latest_release = get_latest_release_version_sync()?;
+    let latest_version = latest_release.version()?;
+    let current_version = crate_version!();
 
-                eprintln!();
+    if current_version == latest_version {
+        opts.terminal
+            .stdout()
+            .plain(fmt_info!(
+                "You are already running the latest version of ockam: {}",
+                current_version
+            ))
+            .write_line()?;
+        return Ok(());
+    }
+
+    if cmd.check {
+        opts.terminal
+            .stdout()
+            .plain(fmt_info!(
+                "A new version of ockam is available: {}. Current ockam version: {}",
+                latest_version,
+                current_version
+            ))
+            .write_line()?;
+        return Ok(());
+    }
+    opts.terminal.write_line(fmt_info!(
+        "A new version of ockam is available: {}. Current ockam version: {}",
+        latest_version,
+        current_version
+    ))?;
+    if !cmd.yes {
+        match opts.terminal.confirm(&fmt_info!(
+            "This will upgrade ockam to the latest version. Are you sure?"
+        ))? {
+            ConfirmResult::Yes => {}
+            ConfirmResult::No => {
+                return Ok(());
+            }
+            ConfirmResult::NonTTY => {
+                return Err(miette!("Use --yes to confirm"));
             }
         }
     }
+    opts.terminal.write_line(fmt_info!(
+        "Upgrading ockam from {} to {}",
+        current_version,
+        latest_version
+    ))?;
+
+    upgrade_ockam(&opts)?;
+    opts.terminal
+        .stdout()
+        .plain(fmt_ok!("Ockam upgraded to version {}", latest_version))
+        .write_line()?;
+    Ok(())
 }
 
-fn upgrade_check_is_disabled() -> bool {
-    get_env_with_default("OCKAM_DISABLE_UPGRADE_CHECK", false).unwrap_or(false)
+fn stop_all_running_nodes(opts: &CommandGlobalOpts) -> Result<Vec<String>> {
+    opts.terminal
+        .write_line(fmt_info!("Stopping all running nodes"))?;
+    let nodes_states = opts.state.nodes.list()?;
+    let mut stopped_nodes = Vec::new();
+    for node_state in nodes_states.iter() {
+        if node_state.is_running() {
+            node_state.kill_process(false)?;
+            opts.terminal
+                .write_line(fmt_ok!("Stopped node {}", node_state.name()))?;
+            stopped_nodes.push(node_state.name().to_string());
+        }
+    }
+    Ok(stopped_nodes)
+}
+
+fn start_nodes(stopped_nodes_names: &[String], opts: &CommandGlobalOpts) -> miette::Result<()> {
+    opts.terminal
+        .write_line(fmt_info!("Restarting all stopped nodes"))?;
+    for node_name in stopped_nodes_names.iter() {
+        let node_state = opts.state.nodes.get(node_name)?;
+        node_state.kill_process(false)?;
```

Comment:
This was part of the original requirements mentioned on the issue here https://github.com/build-trust/ockam/issues/4916.

Reply:
That's not how I read it. Are referring to this line?
>  for now this can restart all the nodes that were stopped.

We have already killed the process of each node, it doesn't seem necessary to do it again.

Reply:
Ah sorry I misunderstood what you meant.
I think that's a relic from earlier attempts where I was having a problem with the nodes not restarting correctly. I'll remove it